controllers/nstemplateset/nstemplateset_controller_test.go

```diff
@@ -1185,6 +1201,14 @@ func toStructured(obj client.Object, decoder runtime.Decoder) (client.Object, er
 			idler := &toolchainv1alpha1.Idler{}
 			_, _, err = decoder.Decode(data, nil, idler)
 			return idler, err
+		case "Role":
+			rl := &rbacv1.Role{}
+			_, _, err = decoder.Decode(data, nil, rl)
+			return rl, err
+		case "RoleBinding":
+			rolebinding := &rbacv1.RoleBinding{}
+			_, _, err = decoder.Decode(data, nil, rolebinding)
+			return rolebinding, err
```

Comment:
Maybe I missed that but do you actually have a unit test(s) which verifies that the role and role binding is restored when deleted?

Reply:
I covered that in the e2e tests, not the unit tests.

Reply:
We usually cover functionality by both unit tests and e2e tests unless there is a good reason to cover it only by e2e tests or in unit tests.
There are multiple reasons for that but one of them is that not everything can be covered by e2e tests. For example error handling is very hard to cover by e2e tests but easy to cover in unit tests.

Reply:
done in 94a94e3